### PR TITLE
Add FCIS training

### PR DESCRIPTION
**Pull request by knorth55**
## Main Changes
- [x] add `ProposalTargetCreator` for `FCIS`
- [x] add `TestProposalTargetCreator`
- [x] add `FCISTrainChain`
- [x] add `TestFCISTrainChain`
- [x] add `tain.py`
- [x] add `train_multi.py`

## Small fixes
- [x] import `experimental.links`
~~- [x] fix `mask_to_bbox`~~ (move to #625 )

## Documentation
- [x] add docstring for `FCISTrainChain`
- [x] add docstring for `ProposalTargetCreator`
- [x] add `FCISTrainChain` in doc
- [x] add `ProposalTargetCreator` in doc

## README
- [x] add `examples/fcis/README.md`
- [x] add `examples/instance_segmentation/README.md`

### CODE REVIEW

**Comment by knorth55** [comment]
Im so sorry for the lateness.
It is ready to be reviewed.

`mAP@0.5: 0.642493`
```
mAP: 0.642493
aeroplane: 0.771879
bicycle: 0.691302
bird: 0.697651
boat: 0.469959
bottle: 0.449674
bus: 0.785370
car: 0.683918
cat: 0.844458
chair: 0.416293
cow: 0.653458
diningtable: 0.383158
dog: 0.808370
horse: 0.744507
motorbike: 0.732374
person: 0.735882
pottedplant: 0.394450
sheep: 0.665110
sofa: 0.498896
train: 0.777762
tvmonitor: 0.645393
```
`mAP@0.7: 0.514390`
```
mAP: 0.514390
aeroplane: 0.653160
bicycle: 0.468946
bird: 0.598293
boat: 0.349783
bottle: 0.355304
bus: 0.741855
car: 0.565675
cat: 0.760414
chair: 0.259050
cow: 0.495735
diningtable: 0.240034
dog: 0.727652
horse: 0.533204
motorbike: 0.561368
person: 0.557394
pottedplant: 0.269213
sheep: 0.533139
sofa: 0.375379
train: 0.682115
tvmonitor: 0.560082
```

File: chainercv/experimental/links/model/fcis/fcis_resnet101.py [hunk 1]
```diff
@@ -248,7 +286,7 @@ def __init__(
                 1024, group_size * group_size * 2 * 4,
                 1, 1, 0, initialW=initialW)
 
-    def __call__(self, x, rois, roi_indices, img_size):
+    def __call__(self, x, rois, roi_indices, img_size, gt_roi_label=None):
```
Comment (yuyu2172) [inline chainercv/experimental/links/model/fcis/fcis_resnet101.py]: `gt_roi_labels (?)`

File: chainercv/experimental/links/model/fcis/fcis_train_chain.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        if cuda.get_array_module(rpn_loc.array) != np:
+            gt_rpn_loc = cuda.to_gpu(gt_rpn_loc)
+            gt_rpn_label = cuda.to_gpu(gt_rpn_label)
+
+        rpn_loc_loss = _fast_rcnn_loc_loss(
+            rpn_loc, gt_rpn_loc, gt_rpn_label, self.rpn_sigma)
+        rpn_cls_loss = F.softmax_cross_entropy(rpn_score, gt_rpn_label)
+
+        # Losses for outputs of the head
+        n_rois = roi_ag_loc.shape[0]
```
Comment (yuyu2172) [inline chainercv/experimental/links/model/fcis/fcis_train_chain.py]: `n_roi`

File: examples/fcis/train.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        'main/roi_cls_loss',
+        'main/roi_mask_loss',
+        'validation/main/map',
+    ]), trigger=print_interval)
+    trainer.extend(
+        extensions.ProgressBar(update_interval=10))
+
+    if extensions.PlotReport.available():
+        import matplotlib
+        matplotlib.use('Agg')
```
Comment (yuyu2172) [inline examples/fcis/train.py]: Can you not set matplotlib setting inside a script?

(related, look at the end of this section 
 https://github.com/chainer/chainercv/tree/master/examples/faster_rcnn#train-code)

File: examples/fcis/train_multi.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    train_iter = chainer.iterators.SerialIterator(train_dataset, batch_size=1)
+
+    if comm.rank == 0:
+        test_dataset = SBDInstanceSegmentationDataset(split='val')
+        test_iter = chainer.iterators.SerialIterator(
+            test_dataset, batch_size=1, repeat=False, shuffle=False)
+
+    # optimizer
+    optimizer = chainermn.create_multi_node_optimizer(
+        chainer.optimizers.MomentumSGD(lr=args.lr * comm.size, momentum=0.9),
```
Comment (yuyu2172) [inline examples/fcis/train_multi.py]: `args.lr * comm.size --> args.lr`

The actual `lr` should be equal to what is passed from the command line.
It gets harder to debug when linear scaling rule is implicitly encoded in a training script.

File: examples/fcis/train_multi.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            'main/roi_loc_loss',
+            'main/roi_cls_loss',
+            'main/roi_mask_loss',
+            'validation/main/map',
+        ]), trigger=print_interval)
+        trainer.extend(
+            extensions.ProgressBar(update_interval=10))
+
+        if extensions.PlotReport.available():
+            import matplotlib
```
Comment (yuyu2172) [inline examples/fcis/train_multi.py]: ditto

File: examples/fcis/train_multi.py (hunk 4 above)
Comment (knorth55) [reply examples/fcis/train_multi.py]: is there any solution to stop gradient mean in `ChainerMN`?

Comment (yuyu2172) [reply examples/fcis/train_multi.py]: What do you mean by `stop gradient mean`?

Comment (knorth55) [reply examples/fcis/train_multi.py]: chainermn default is gather gradient, average it and backpropagate it to all model. 
i want to just skip the mean phase.

Comment (yuyu2172) [reply examples/fcis/train_multi.py]: Unfortunately, I don't know how to that.

I like to keep loss divided by batchsize always, so I never tried doing it.

Comment (knorth55) [reply examples/fcis/train_multi.py]: I updated and add help.

**Comment by yuyu2172** [comment]
Can I merge?

**Comment by knorth55** [comment]
Yes, please

**Comment by yuyu2172** [comment]
LGTM!
